internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +115,328 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+				Message: `spec.githubAPI.host ("") is not valid`,
+			},
+			tlsConfigurationUnknown,
+			githubConnectionUnknown,
+		}, "", nil, nil
+	}
+
+	host := *gitHubAPIConfig.Host
+	hostPort, addressParseErr := endpointaddr.Parse(host, 443)
+	if addressParseErr != nil {
+		// addressParseErr is not recoverable. It requires user interaction, so do not return the error.
+		return []*metav1.Condition{
+			{
+				Type:    HostValid,
+				Status:  metav1.ConditionFalse,
+				Reason:  "InvalidHost",
+				Message: fmt.Sprintf("spec.githubAPI.host (%q) is not valid", host),
+			},
+			tlsConfigurationUnknown,
+			githubConnectionUnknown,
+		}, host, nil, nil
+	}
+
+	hostOnly := *gitHubAPIConfig.Host
+	hostWithHttps := fmt.Sprintf("https://%s", hostOnly)
+	conditions := make([]*metav1.Condition, 0)
+	conditions = append(conditions, &metav1.Condition{
+		Type:    HostValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: fmt.Sprintf("spec.githubAPI.host (%q) is valid", host),
+	})
+
+	certPool, _, buildCertPoolErr := buildCertPool(gitHubAPIConfig.TLS)
+	if buildCertPoolErr != nil {
+		// buildCertPoolErr is not recoverable. It requires user interaction, so do not return the error.
+		conditions = append(conditions, &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  "InvalidTLSConfig",
+			Message: fmt.Sprintf("tls configuration is not valid for spec.githubAPI.host (%q): %s", host, buildCertPoolErr),
+		})
+		conditions = append(conditions, githubConnectionUnknown)
+
+		return conditions, hostWithHttps, nil, nil
+	}
+
+	tlsDialer := tls.Dialer{
+		NetDialer: &net.Dialer{
+			Timeout: 30 * time.Second,
+		},
+		Config: &tls.Config{
+			MinVersion: tls.VersionTLS12,
+			// If certPool is nil then RootCAs will be set to nil and TLS will use the host's root CA set automatically.
+			RootCAs: certPool,
+		},
+	}
+
+	conn, tlsDialErr := tlsDialer.DialContext(ctx, "tcp", hostPort.Endpoint())
+	if tlsDialErr != nil {
+		conditions = append(conditions, &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  upstreamwatchers.ReasonInvalidTLSConfig,
+			Message: fmt.Sprintf("tls configuration is not valid for spec.githubAPI.host (%q): %s", host, tlsDialErr.Error()),
+		})
+		conditions = append(conditions, githubConnectionUnknown)
+		return conditions, hostWithHttps, nil, tlsDialErr
+	}
+
+	conditions = append(conditions, &metav1.Condition{
+		Type:    TLSConfigurationValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: fmt.Sprintf("tls configuration is valid for spec.githubAPI.host (%q)", host),
+	})
+	conditions = append(conditions, &metav1.Condition{
+		Type:    GitHubConnectionValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: fmt.Sprintf("spec.githubAPI.host (%q) is reachable and TLS verification succeeds", host),
+	})
+
```

Comment:
I wonder if we care if the `conn.Close()` errors or not?  Per https://github.com/vmware-tanzu/pinniped/blob/main/internal/controller/authenticator/webhookcachefiller/webhookcachefiller.go#L304C1-L307C3.

Reply:
It doesn't feel right to just completely ignore the error, so we'll treat it as an application error (not user error) which should trigger a resync.